In [1]:
import sys
sys.path.append("..")

from src.data_preprocessing import build_preprocessor
from src.imbalance_handler import apply_smote

In [4]:
import pandas as pd

df = pd.read_csv("../data/processed/fraud_processed.csv")
X = df.drop("class", axis=1)
y = df["class"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

In [5]:
# THEN preprocessing
from src.data_preprocessing import build_preprocessor

preprocessor = build_preprocessor(X_train)

X_train_scaled = preprocessor.fit_transform(X_train)
X_test_scaled = preprocessor.transform(X_test)

In [6]:
# THEN SMOTE (ONLY on training set)
from src.imbalance_handler import apply_smote

X_train_balanced, y_train_balanced = apply_smote(
    X_train_scaled,
    y_train
)

Before SMOTE
class
0    93502
1     9814
Name: count, dtype: int64
After SMOTE
class
0    93502
1    93502
Name: count, dtype: int64


# Class Imbalance Handling (SMOTE) – Interpretation
### Before SMOTE

-  The dataset shows a severe class imbalance between legitimate and fraud transactions:
| Class          | Count  |
| -------------- | ------ |
| 0 (Legitimate) | 93,502 |
| 1 (Fraud)      | 9,814  |

-  This indicates that fraud cases represent only a small fraction of the dataset, making the problem highly imbalanced.
### Key Observation:
-  Fraud cases ≈ 9.5% of the dataset
-  Legitimate transactions dominate the dataset
-  A model trained on this data without balancing would likely be biased toward predicting class 0
### Business Impact of Imbalance
-  The model may achieve high accuracy by predicting only legitimate transactions
-  However, it would fail to detect fraud effectively
-  This leads to high financial risk due to false negatives (missed fraud cases)
### After SMOTE (Synthetic Minority Oversampling Technique)

-  After applying SMOTE, the dataset becomes perfectly balanced:
| Class          | Count  |
| -------------- | ------ |
| 0 (Legitimate) | 93,502 |
| 1 (Fraud)      | 93,502 |

### Key Observation:
-  Fraud class has been synthetically oversampled to match the majority class
-  The dataset is now fully balanced (1:1 ratio)
### Why This Matters

SMOTE improves model learning by:

-  Preventing bias toward the majority class
-  Allowing the model to learn fraud patterns more effectively
-  Improving recall and F1-score for fraud detection
- Reducing the risk of missing fraudulent transactions
### Final Justification

#### SMOTE was selected over undersampling because:

-  Undersampling would discard a large number of legitimate transactions (loss of information)
-  SMOTE preserves all original data while generating synthetic fraud samples
-  This leads to a more robust and generalizable fraud detection model
### Conclusion

The resampling process successfully transformed a highly imbalanced dataset into a balanced one, which is essential for building an effective fraud detection system capable of identifying rare but critical fraudulent transactions.